internal/app/client/client.go

```diff
@@ -0,0 +1,164 @@
+// Copyright 2023 The Connect Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package client
+
+import (
+	"context"
+	"crypto/tls"
+	"errors"
+	"flag"
+	"io"
+	"net"
+	"net/http"
+	"net/url"
+	"strconv"
+
+	"connectrpc.com/conformance/internal/app"
+	"connectrpc.com/conformance/internal/gen/proto/connect/connectrpc/conformance/v1alpha1/conformancev1alpha1connect"
+	v1alpha1 "connectrpc.com/conformance/internal/gen/proto/go/connectrpc/conformance/v1alpha1"
+	"connectrpc.com/connect"
+	"golang.org/x/net/http2"
+)
+
+// Run runs the client according to a client config read from the 'in' reader. The result of the run
+// is written to the 'out' writer, including any errors encountered during the actual run. Any error
+// returned from this function is indicative of an issue with the reader or writer and should not be related
+// to the actual run.
+func Run(ctx context.Context, _ []string, inReader io.ReadCloser, outWriter io.WriteCloser) error {
+	json := flag.Bool("json", false, "whether to use the JSON format for marshaling / unmarshaling messages")
+
+	flag.Parse()
+
+	// Read the server config from  the in reader
+	data, err := io.ReadAll(inReader)
+	if err != nil {
+		return err
+	}
```

Comment:
Sorry I didn't catch this before. And double-sorry since it was not spelled out anywhere in any of the comments (though I think may have been alluded to in a comment in the design in Notion).

But the intent here was that a client might read _many_ RPC definitions. We don't want the overhead of spawning a new process for every single request.

The format was to be varint-length-prefixed protos.

So the client should be reading input in a loop, something like so:
```go
// pseudo-code-ish
for {
    // read varint-encoded prefix that indicates message length
    length, err := readVarintLength(inReader)
    if errors.Is(err, io.EOF) { return nil /* all done */ }
    if err != nil { return err }
    // Read the message
    data := make([]byte, length)
    if _, err := io.ReadFull(inReader, data); err != nil {
        return err
    }

    // ... Now we can unmarshal the request, issue the RPC,
    //     write the response results, and then go to top of
    //     loop to read another request ...
}
```

This was the side discussion where Timo asked about the format here: https://www.notion.so/Connect-Conformance-Tests-Design-0a279279a89c48559453ac3c7474b063?d=3e6bb2676deb43389dfa0546bb760ac6&pvs=4#4f3088105113406cbf8c371dda1ac999

And now that I think about, "varint-encoded" prefix is unnecessarily complicated. Instead, it could just read four bytes as network-byte-order `uint32` for the size, and then read that many more bytes for the actual message. I say varint is too complicated because it looks like there isn't actually any decent implementation to do a streaming read of a varint-encoded number from an `io.Reader`. So maybe it would just be easier to make it a fixed-length integer...

FWIW, I'm not sure how this would work with JSON inputs. Maybe we can keep this as is for now and just add a TODO, and then we can follow up and make sure that whatever format this reads is the same as the format that the test runner actually writes.

(Aside: it is only necessary for the client to act this way. A server will only ever read/write a single protobuf message, not a stream of them, so we should leave it as is.)

Reply:
Sounds good. I added a TODO and we can address it when the runner is ready for primetime.